protocols/request-response/src/handler.rs

```diff
@@ -104,57 +107,144 @@ where
             keep_alive: KeepAlive::Yes,
             keep_alive_timeout,
             substream_timeout,
-            outbound: VecDeque::new(),
+            pending_outbound: VecDeque::new(),
+            requested_outbound: Default::default(),
             inbound: FuturesUnordered::new(),
             pending_events: VecDeque::new(),
-            pending_error: None,
             inbound_request_id,
+            worker_streams: Default::default(),
         }
     }
 
     fn on_fully_negotiated_inbound(
         &mut self,
         FullyNegotiatedInbound {
-            protocol: sent,
-            info: request_id,
+            protocol: (mut stream, protocol),
+            info: (),
         }: FullyNegotiatedInbound<
             <Self as ConnectionHandler>::InboundProtocol,
             <Self as ConnectionHandler>::InboundOpenInfo,
         >,
     ) {
-        if sent {
-            self.pending_events
-                .push_back(Event::ResponseSent(request_id))
-        } else {
-            self.pending_events
-                .push_back(Event::ResponseOmission(request_id))
-        }
+        let mut codec = self.codec.clone();
+
+        // A channel for notifying the handler when the inbound
+        // upgrade received the request.
+        let (rq_send, rq_recv) = oneshot::channel();
+
+        // A channel for notifying the inbound upgrade when the
+        // response is sent.
+        let (rs_send, rs_recv) = oneshot::channel();
+
+        // The handler waits for the request to come in. It then emits
+        // `Event::Request` together with a
+        // `ResponseChannel`.
+        self.inbound
+            .push(rq_recv.map_ok(move |rq| (rq, rs_send)).boxed());
+
+        let request_id = RequestId(self.inbound_request_id.fetch_add(1, Ordering::Relaxed));
+        let timeout = self.substream_timeout;
+
+        let recv = async move {
+            let read = codec.read_request(&protocol, &mut stream);
+            let request = read.await?;
```

Comment:
```suggestion
            let request = codec.read_request(&protocol, &mut stream).await?;
```

Wouldn't that compile as well?

Reply:
Probably, I just verbatim copied what the previous code did.

Reply:
Causes some kind of `Send`/`Sync` error with `protocol`.